File: domain/src/main/java/com/crhonvas/domain/model/user/Address.java
package com.crhonvas.domain.model.user;

import androidx.annotation.NonNull;

import org.apache.commons.lang3.builder.EqualsBuilder;
import org.apache.commons.lang3.builder.HashCodeBuilder;

public class Address {
    private String street;
    private String suite;
    private String city;
    private String zipCode;
    private Geo geo;

    private Address(@NonNull Builder builder) {
        this.street = builder.street;
        this.suite = builder.suite;
        this.city = builder.city;
        this.zipCode = builder.zipCode;
        this.geo = builder.geo;
    }

    @NonNull
    public String getStreet() {
        return street == null ? "" : street;
    }

    @NonNull
    public String getSuite() {
        return suite == null ? "" : suite;
    }

    @NonNull
    public String getCity() {
        return city == null ? "" : city;
    }

    @NonNull
    public String getZipCode() {
        return zipCode == null ? "" : zipCode;
    }

    @NonNull
    public Geo getGeo() {
        return geo == null ? new Geo.Builder().build() : geo;
    }

    @Override
    public boolean equals(Object o) {
        if (this == o) return true;

        if (o == null || getClass() != o.getClass()) return false;

        Address address = (Address) o;

        return new EqualsBuilder()
                .append(street, address.street)
                .append(suite, address.suite)
                .append(city, address.city)
                .append(zipCode, address.zipCode)
                .append(geo, address.geo)
                .isEquals();
    }

    @Override
    public int hashCode() {
        return new HashCodeBuilder(17, 37)
                .append(street)
                .append(suite)
                .append(city)
                .append(zipCode)
                .append(geo)
                .toHashCode();
    }

    public static final class Builder {
        private String street;
        private String suite;
        private String city;
        private String zipCode;
        private Geo geo;

        public Builder() {}

        @NonNull
        public Builder street(@NonNull String street) {
            this.street = street;
            return this;
        }

        @NonNull
        public Builder suite(@NonNull String suite) {
            this.suite = suite;
            return this;
        }

        @NonNull
        public Builder city(@NonNull String city) {
            this.city = city;
            return this;
        }

        @NonNull
        public Builder zipCode(@NonNull String zipCode) {
            this.zipCode = zipCode;
            return this;
        }

        @NonNull
        public Builder geo(@NonNull Geo geo) {
            this.geo = geo;
            return this;
        }

        @NonNull
        public Address build() {
            return new Address(this);
        }
    }
}

File: domain/src/main/java/com/crhonvas/domain/model/user/Address.kt
package com.crhonvas.domain.model.user

import org.apache.commons.lang3.builder.EqualsBuilder
import org.apache.commons.lang3.builder.HashCodeBuilder

class Address private constructor(builder: Builder) {
    private val street: String?
    private val suite: String?
    private val city: String?
    private val zipCode: String?
    private val geo: Geo?

    init {
        street = builder.street
        suite = builder.suite
        city = builder.city
        zipCode = builder.zipCode
        geo = builder.geo
    }

    fun getStreet(): String {
        return street ?: ""
    }

    fun getSuite(): String {
        return suite ?: ""
    }

    fun getCity(): String {
        return city ?: ""
    }

    fun getZipCode(): String {
        return zipCode ?: ""
    }

    fun getGeo(): Geo {
        return geo ?: Geo.Builder().build()
    }

    override fun equals(o: Any?): Boolean {
        if (this === o) return true
        if (o == null || javaClass != o.javaClass) return false
        val address = o as Address
        return EqualsBuilder()
            .append(street, address.street)
            .append(suite, address.suite)
            .append(city, address.city)
            .append(zipCode, address.zipCode)
            .append(geo, address.geo)
            .isEquals
    }

    override fun hashCode(): Int {
        return HashCodeBuilder(17, 37)
            .append(street)
            .append(suite)
            .append(city)
            .append(zipCode)
            .append(geo)
            .toHashCode()
    }

    class Builder {
        var street: String? = null
        var suite: String? = null
        var city: String? = null
        var zipCode: String? = null
        var geo: Geo? = null
        fun street(street: String): Builder {
            this.street = street
            return this
        }

        fun suite(suite: String): Builder {
            this.suite = suite
            return this
        }

        fun city(city: String): Builder {
            this.city = city
            return this
        }

        fun zipCode(zipCode: String): Builder {
            this.zipCode = zipCode
            return this
        }

        fun geo(geo: Geo): Builder {
            this.geo = geo
            return this
        }

        fun build(): Address {
            return Address(this)
        }
    }
}